# Sprint2. Analistes del Perfil del Client: 
#### Com podem ajustar les nostres ofertes i promocions per alinearnos millor amb les característiques demogràfiques dels nostres clients més valuosos?

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
df_BANK_marketing = pd.read_csv('df_BANK_marketing_Sprint2.csv')

In [3]:
df_BANK_marketing

id   age           job   marital  education  is_in_default  balance  \
0          1  59.0        admin.   married  secondary              0     2343   
1          2  59.0        admin.   married  secondary              0     2343   
2          3  56.0        admin.   married  secondary              0       45   
3          4  41.0    technician   married  secondary              0     1270   
4          5  55.0      services   married  secondary              0     2476   
...      ...   ...           ...       ...        ...            ...      ...   
16158  21379  80.0       retired   married  secondary              0     8304   
16159  21380  30.0    unemployed  divorced  secondary              0     7105   
16160  21381  33.0  entrepreneur    single  secondary              0      278   
16161  21382  57.0    technician   married    primary              0     3376   
16162  21383  28.0    management    single   tertiary              0       41   

       housing  loan    contact  day month  duration  campaign  pdays  \
0            1     0    unknown    5   may      1042         1     -1   
1            1     0    unknown    5   may      1042         1     -1   
2            0     0    unknown    5   may      1467         1     -1   
3            1     0    unknown    5   may      1389         1     -1   
4            1     0    unknown    5   may       579         1     -1   
...        ...   ...        ...  ...   ...       ...       ...    ...   
16158        0     0  telephone    9   jun       712         1     64   
16159        0     0   cellular   22   dec       321         2     -1   
16160        0     0   cellular    1   feb       563         5     -1   
16161        1     0  telephone    2   jun       421         2     -1   
16162        1     0   cellular   15   may       885         3    352   

       previous poutcome  deposit  balance_normalizado    estado_balance  
0             0  unknown        1             0.246103        Saldo Alto  
1             0  unknown        1             0.246103        Saldo Alto  
2             0  unknown        1            -0.445637        Saldo Bajo  
3             0  unknown        1            -0.076890  Saldo Medio Alto  
4             0  unknown        1             0.286138        Saldo Alto  
...         ...      ...      ...                  ...               ...  
16158        12  failure        1             2.040471    Saldo Muy Alto  
16159         0  unknown        1             1.679550    Saldo Muy Alto  
16160         0  unknown        1            -0.375500       Saldo Medio  
16161         0  unknown        1             0.557054        Saldo Alto  
16162         2    other        1            -0.446841        Saldo Bajo  

[16163 rows x 20 columns]

In [4]:
cliente= df_BANK_marketing[['id', 'age', 'job', 'marital', 'education','balance', 'is_in_default', 'housing', 'loan', 'deposit']]
cliente

id   age           job   marital  education  balance  is_in_default  \
0          1  59.0        admin.   married  secondary     2343              0   
1          2  59.0        admin.   married  secondary     2343              0   
2          3  56.0        admin.   married  secondary       45              0   
3          4  41.0    technician   married  secondary     1270              0   
4          5  55.0      services   married  secondary     2476              0   
...      ...   ...           ...       ...        ...      ...            ...   
16158  21379  80.0       retired   married  secondary     8304              0   
16159  21380  30.0    unemployed  divorced  secondary     7105              0   
16160  21381  33.0  entrepreneur    single  secondary      278              0   
16161  21382  57.0    technician   married    primary     3376              0   
16162  21383  28.0    management    single   tertiary       41              0   

       housing  loan  deposit  
0            1     0        1  
1            1     0        1  
2            0     0        1  
3            1     0        1  
4            1     0        1  
...        ...   ...      ...  
16158        0     0        1  
16159        0     0        1  
16160        0     0        1  
16161        1     0        1  
16162        1     0        1  

[16163 rows x 10 columns]

### APUNTES ESTADISTICOS
Variable resultado numérica: Usa una matriz de correlación para ver la relación con otras variables numéricas.
Variable resultado binaria y características categóricas: Aplica la prueba de Chi-cuadrado para evaluar relaciones significativas.
Variable resultado binaria y características numéricas: Utiliza una prueba t o ANOVA para comparar las medias de las variables numéricas entre las dos categorías de la variable binaria.

### HIPOTECA
### Influencia de variables en la contratación de hipotecas

### VARIABLE TRABAJO

In [5]:
# Variable numérica y variable binaria
# Example of the Student's t-test
from scipy.stats import ttest_ind
data1 = cliente['housing']
data2 = cliente['age']
stat, p = ttest_ind(data1, data2)
print('stat=%.3f, p=%.3f' % (stat, p))
if p > 0.05:
	print('Probably the same distribution')
else:
	print('Probably different distributions')

## por el resultado no hay una influencia de la edad en la contratación de hipóteca

stat=-434.214, p=0.000
Probably different distributions


### VARIABLE JOB

In [6]:
tabla_job_housing = pd.crosstab(cliente['housing'], cliente['job'])

In [7]:
tabla_job_housing

job      admin.  blue-collar  entrepreneur  housemaid  management  retired  \
housing                                                                      
0           857          911           228        291        2174      953   
1          1074         1903           232        112        1545      186   

job      self-employed  services  student  technician  unemployed  unknown  
housing                                                                     
0                  337       536      449        1363         340      104  
1                  239       820       76        1260         167        6

In [8]:
# Variable categórica y variable binaria
# Example of the Chi-Squared Test
from scipy.stats import chi2_contingency
stat, p, dof, expected = chi2_contingency(tabla_job_housing)
print('stat=%.3f, p=%.3f' % (stat, p))
if p > 0.05:
	print('Probably independent')
else:
	print('Probably dependent')

## Según el resultado del test, existe una relación entre los diferentes trabajos y que tengan o no hipóteca

stat=1520.306, p=0.000
Probably dependent


### VARIABLE MARITAL

In [9]:
tabla_marital_housing = pd.crosstab(cliente['housing'], cliente['marital'])
tabla_marital_housing

marital  divorced  married  single
housing                           
0             981     4702    2860
1             900     4470    2250

In [10]:
# Variable categórica y variable binaria
# Example of the Chi-Squared Test
from scipy.stats import chi2_contingency
stat, p, dof, expected = chi2_contingency(tabla_marital_housing)
print('stat=%.3f, p=%.3f' % (stat, p))
if p > 0.05:
	print('Probably independent')
else:
	print('Probably dependent')
	
## Según el resultado del test, existe una relación entre los diferentes estados civiles y que tengan o no hipóteca

stat=29.562, p=0.000
Probably dependent


### VARIABLE EDUCATION

In [11]:
tabla_education_housing = pd.crosstab(cliente['housing'], cliente['education'])
tabla_education_housing

education  primary  secondary  tertiary  unknown
housing                                         
0             1108       3727      3237      471
1             1056       4216      2101      247

In [12]:
# Variable categórica y variable binaria
# Example of the Chi-Squared Test
from scipy.stats import chi2_contingency
stat, p, dof, expected = chi2_contingency(tabla_marital_housing)
print('stat=%.3f, p=%.3f' % (stat, p))
if p > 0.05:
	print('Probably independent')
else:
	print('Probably dependent')
	
## Según el resultado del test, existe una relación entre los diferentes estudios y que tengan o no hipóteca

stat=29.562, p=0.000
Probably dependent


### VARIABLE is_in_default

In [13]:
tabla_impagos_housing = pd.crosstab(cliente['housing'], cliente['is_in_default'])
tabla_impagos_housing

is_in_default     0    1
housing                 
0              8422  121
1              7492  128

In [14]:
#  2 variableS binarias (dos variables categóricas)
# Example of the Chi-Squared Test
from scipy.stats import chi2_contingency
stat, p, dof, expected = chi2_contingency(tabla_impagos_housing)
print('stat=%.3f, p=%.3f' % (stat, p))
if p > 0.05:
	print('Probably independent')
else:
	print('Probably dependent')
	
## Según el resultado del test, no existe una relación entre los diferentes los impagos y que tengan o no hipóteca

stat=1.673, p=0.196
Probably independent


### VARIABLE BALANCE

In [15]:
# Variable numérica y variable binaria
# Example of the Student's t-test
from scipy.stats import ttest_ind
data1 = cliente['housing']
data2 = cliente['balance']
stat, p = ttest_ind(data1, data2)
print('stat=%.3f, p=%.3f' % (stat, p))
if p > 0.05:
	print('Probably the same distribution')
else:
	print('Probably different distributions')

## por el resultado no hay una influencia del balance en la contratación de hipóteca

stat=-58.358, p=0.000
Probably different distributions


### ANALISIS MULTIVARIANTE

In [16]:
from statsmodels.formula.api import ols
from statsmodels.stats.anova import anova_lm

# Convertir variable binaria a proporciones por grupo
cliente['housing_mean'] = cliente.groupby(['job', 'marital', 'education'])['housing'].transform('mean')

# Modelo ANOVA
modelo = ols('housing_mean ~ C(job) + C(marital) + C(education)', data=cliente).fit()
anova_results = anova_lm(modelo)

print(anova_results)

                   df      sum_sq    mean_sq            F         PR(>F)
C(job)           11.0  378.837045  34.439731  7212.603159   0.000000e+00
C(marital)        2.0    5.529263   2.764631   578.987924  1.433750e-243
C(education)      3.0   14.037743   4.679248   979.959928   0.000000e+00
Residual      16146.0   77.096146   0.004775          NaN            NaN


C:\Users\34622\AppData\Local\Temp\ipykernel_10984\757918150.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cliente['housing_mean'] = cliente.groupby(['job', 'marital', 'education'])['housing'].transform('mean')


## LOAN

In [17]:
from statsmodels.formula.api import ols
from statsmodels.stats.anova import anova_lm

# Convertir variable binaria a proporciones por grupo
cliente['loan_mean'] = cliente.groupby(['job', 'marital', 'education'])['loan'].transform('mean')

# Modelo ANOVA
modelo = ols('loan_mean ~ C(job) + C(marital) + C(education)', data=cliente).fit()
anova_results = anova_lm(modelo)

print(anova_results);

                   df     sum_sq   mean_sq            F  PR(>F)
C(job)           11.0  33.573130  3.052103  2606.753806     0.0
C(marital)        2.0   6.806151  3.403075  2906.514153     0.0
C(education)      3.0   6.396334  2.132111  1821.003271     0.0
Residual      16146.0  18.904451  0.001171          NaN     NaN


C:\Users\34622\AppData\Local\Temp\ipykernel_10984\4145049378.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cliente['loan_mean'] = cliente.groupby(['job', 'marital', 'education'])['loan'].transform('mean')


CONCLUSIONES:
1) Todas las variables categóricas (job, marital, education) tienen un efecto significativo sobre la variable dependiente (loan), dado que los p-valores son <0.05
2) La variable job tiene el mayor impacto en la variabilidad explicada (𝐹=2606.75 y suma de cuadrados más alta: 
33.57).
3) Residual mean_sq = 0.001171. La variabilidad residual es pequeña, lo que indica que el modelo ANOVA ajusta bien los datos.

Teoria:
df: número de categórias
sum_sq (Suma de cuadrados):Representa la variabilidad explicada por la variable correspondiente.
F (Estadístico F): Valores altos de 𝐹 sugieren que la variable tiene un efecto significativo.
PR(>F) (p-valor): Evalúa la significancia estadística del efecto de la variable. Si el p-valor es menor a 
0.05 se considera que la variable tiene un impacto significativo.

## DEPOSIT

In [18]:
from statsmodels.formula.api import ols
from statsmodels.stats.anova import anova_lm

# Convertir variable binaria a proporciones por grupo
cliente['deposit_mean'] = cliente.groupby(['job', 'marital', 'education'])['deposit'].transform('mean')

# Modelo ANOVA
modelo = ols('deposit_mean ~ C(job) + C(marital) + C(education)', data=cliente).fit()
anova_results = anova_lm(modelo)

print(anova_results)

C:\Users\34622\AppData\Local\Temp\ipykernel_10984\2780215874.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cliente['deposit_mean'] = cliente.groupby(['job', 'marital', 'education'])['deposit'].transform('mean')


                   df      sum_sq    mean_sq            F  PR(>F)
C(job)           11.0  139.503147  12.682104  3189.494189     0.0
C(marital)        2.0   26.320046  13.160023  3309.688705     0.0
C(education)      3.0   21.909784   7.303261  1836.738565     0.0
Residual      16146.0   64.199915   0.003976          NaN     NaN


### Modelo de regresión logística para variables binarias (housing, loan, deposit)

In [19]:
import statsmodels.api as sm
import patsy

### HOUSING

In [20]:
# Definir la fórmula para el modelo de regresión logística
formula = 'housing ~ C(job) + C(marital) + C(education) + is_in_default + balance'

# Preparar las variables dependientes (y) e independientes (X)
y, X = patsy.dmatrices(formula, cliente, return_type='dataframe')

# Ajustar el modelo logístico
modelo = sm.Logit(y, X).fit()

# Resumen del modelo
print(modelo.summary())

Optimization terminated successfully.
         Current function value: 0.636987
         Iterations 7
                           Logit Regression Results                           
Dep. Variable:                housing   No. Observations:                16163
Model:                          Logit   Df Residuals:                    16144
Method:                           MLE   Df Model:                           18
Date:                Thu, 05 Dec 2024   Pseudo R-squ.:                 0.07885
Time:                        11:22:03   Log-Likelihood:                -10296.
converged:                       True   LL-Null:                       -11177.
Covariance Type:            nonrobust   LLR p-value:                     0.000
                                coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------------
Intercept                     0.3483      0.084      4.126      0.000       0.

### LOAN

In [21]:
# Definir la fórmula para el modelo de regresión logística
formula = 'loan ~ C(job) + C(marital) + C(education) + is_in_default + balance'

# Preparar las variables dependientes (y) e independientes (X)
y, X = patsy.dmatrices(formula, cliente, return_type='dataframe')

# Ajustar el modelo logístico
modelo = sm.Logit(y, X).fit()

# Resumen del modelo
print(modelo.summary())

Optimization terminated successfully.
         Current function value: 0.367733
         Iterations 10
                           Logit Regression Results                           
Dep. Variable:                   loan   No. Observations:                16163
Model:                          Logit   Df Residuals:                    16144
Method:                           MLE   Df Model:                           18
Date:                Thu, 05 Dec 2024   Pseudo R-squ.:                 0.05482
Time:                        11:22:03   Log-Likelihood:                -5943.7
converged:                       True   LL-Null:                       -6288.4
Covariance Type:            nonrobust   LLR p-value:                9.958e-135
                                coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------------
Intercept                    -1.2890      0.112    -11.534      0.000      -1

### DEPOSIT

In [22]:
# Definir la fórmula para el modelo de regresión logística
formula = 'deposit ~ C(job) + C(marital) + C(education) + is_in_default + balance'

# Preparar las variables dependientes (y) e independientes (X)
y, X = patsy.dmatrices(formula, cliente, return_type='dataframe')

# Ajustar el modelo logístico
modelo = sm.Logit(y, X).fit()

# Resumen del modelo
print(modelo.summary())

Optimization terminated successfully.
         Current function value: 0.664510
         Iterations 5
                           Logit Regression Results                           
Dep. Variable:                deposit   No. Observations:                16163
Model:                          Logit   Df Residuals:                    16144
Method:                           MLE   Df Model:                           18
Date:                Thu, 05 Dec 2024   Pseudo R-squ.:                 0.03945
Time:                        11:22:03   Log-Likelihood:                -10740.
converged:                       True   LL-Null:                       -11182.
Covariance Type:            nonrobust   LLR p-value:                9.955e-176
                                coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------------
Intercept                    -0.4109      0.082     -5.004      0.000      -0.

In [23]:
# cliente['group_combination'] = cliente['job'] + '|' + cliente['marital'] + '|' + cliente['education']

In [24]:
cliente['housing'] = cliente['housing'].replace({1:'hipoteca'})
cliente['loan'] = cliente['loan'].replace({1:'credito'})
cliente['deposit'] = cliente['deposit'].replace({1:'deposit'})
cliente

C:\Users\34622\AppData\Local\Temp\ipykernel_10984\2221701948.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cliente['housing'] = cliente['housing'].replace({1:'hipoteca'})
C:\Users\34622\AppData\Local\Temp\ipykernel_10984\2221701948.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cliente['loan'] = cliente['loan'].replace({1:'credito'})
C:\Users\34622\AppData\Local\Temp\ipykernel_10984\2221701948.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try us

id   age           job   marital  education  balance  is_in_default  \
0          1  59.0        admin.   married  secondary     2343              0   
1          2  59.0        admin.   married  secondary     2343              0   
2          3  56.0        admin.   married  secondary       45              0   
3          4  41.0    technician   married  secondary     1270              0   
4          5  55.0      services   married  secondary     2476              0   
...      ...   ...           ...       ...        ...      ...            ...   
16158  21379  80.0       retired   married  secondary     8304              0   
16159  21380  30.0    unemployed  divorced  secondary     7105              0   
16160  21381  33.0  entrepreneur    single  secondary      278              0   
16161  21382  57.0    technician   married    primary     3376              0   
16162  21383  28.0    management    single   tertiary       41              0   

        housing loan  deposit  housing_mean  loan_mean  deposit_mean  
0      hipoteca    0  deposit      0.553191   0.158949      0.443054  
1      hipoteca    0  deposit      0.553191   0.158949      0.443054  
2             0    0  deposit      0.553191   0.158949      0.443054  
3      hipoteca    0  deposit      0.507609   0.163043      0.431522  
4      hipoteca    0  deposit      0.670588   0.194958      0.349580  
...         ...  ...      ...           ...        ...           ...  
16158         0    0  deposit      0.177112   0.084469      0.656676  
16159         0    0  deposit      0.385965   0.052632      0.561404  
16160         0    0  deposit      0.433333   0.133333      0.600000  
16161  hipoteca    0  deposit      0.531915   0.127660      0.297872  
16162  hipoteca    0  deposit      0.382562   0.067616      0.574733  

[16163 rows x 13 columns]

In [25]:
cliente['housing'] = cliente['housing'].astype(str)
cliente['loan'] = cliente['loan'].astype(str)
cliente['deposit'] = cliente['deposit'].astype(str)

C:\Users\34622\AppData\Local\Temp\ipykernel_10984\318873528.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cliente['housing'] = cliente['housing'].astype(str)
C:\Users\34622\AppData\Local\Temp\ipykernel_10984\318873528.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cliente['loan'] = cliente['loan'].astype(str)
C:\Users\34622\AppData\Local\Temp\ipykernel_10984\318873528.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_i

In [26]:
cliente['perfil'] = cliente['job'] + '_' + cliente['education'] + '_' + cliente['marital']
cliente['producto'] = cliente['housing'] + '_' + cliente['loan'] + '_' + cliente['deposit']
cliente

C:\Users\34622\AppData\Local\Temp\ipykernel_10984\3859738216.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cliente['perfil'] = cliente['job'] + '_' + cliente['education'] + '_' + cliente['marital']
C:\Users\34622\AppData\Local\Temp\ipykernel_10984\3859738216.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cliente['producto'] = cliente['housing'] + '_' + cliente['loan'] + '_' + cliente['deposit']


id   age           job   marital  education  balance  is_in_default  \
0          1  59.0        admin.   married  secondary     2343              0   
1          2  59.0        admin.   married  secondary     2343              0   
2          3  56.0        admin.   married  secondary       45              0   
3          4  41.0    technician   married  secondary     1270              0   
4          5  55.0      services   married  secondary     2476              0   
...      ...   ...           ...       ...        ...      ...            ...   
16158  21379  80.0       retired   married  secondary     8304              0   
16159  21380  30.0    unemployed  divorced  secondary     7105              0   
16160  21381  33.0  entrepreneur    single  secondary      278              0   
16161  21382  57.0    technician   married    primary     3376              0   
16162  21383  28.0    management    single   tertiary       41              0   

        housing loan  deposit  housing_mean  loan_mean  deposit_mean  \
0      hipoteca    0  deposit      0.553191   0.158949      0.443054   
1      hipoteca    0  deposit      0.553191   0.158949      0.443054   
2             0    0  deposit      0.553191   0.158949      0.443054   
3      hipoteca    0  deposit      0.507609   0.163043      0.431522   
4      hipoteca    0  deposit      0.670588   0.194958      0.349580   
...         ...  ...      ...           ...        ...           ...   
16158         0    0  deposit      0.177112   0.084469      0.656676   
16159         0    0  deposit      0.385965   0.052632      0.561404   
16160         0    0  deposit      0.433333   0.133333      0.600000   
16161  hipoteca    0  deposit      0.531915   0.127660      0.297872   
16162  hipoteca    0  deposit      0.382562   0.067616      0.574733   

                              perfil            producto  
0           admin._secondary_married  hipoteca_0_deposit  
1           admin._secondary_married  hipoteca_0_deposit  
2           admin._secondary_married         0_0_deposit  
3       technician_secondary_married  hipoteca_0_deposit  
4         services_secondary_married  hipoteca_0_deposit  
...                              ...                 ...  
16158      retired_secondary_married         0_0_deposit  
16159  unemployed_secondary_divorced         0_0_deposit  
16160  entrepreneur_secondary_single         0_0_deposit  
16161     technician_primary_married  hipoteca_0_deposit  
16162     management_tertiary_single  hipoteca_0_deposit  

[16163 rows x 15 columns]

In [27]:
cliente_selec = cliente[['id','age', 'perfil', 'producto', 'is_in_default', 'balance']]
cliente_selec = cliente_selec[cliente_selec['producto'] != '0_0_0']
cliente_selec['producto'] = cliente_selec['producto'].str.replace('0','', regex=False).str.strip('_')
cliente_selec

id   age                         perfil           producto  \
0          1  59.0       admin._secondary_married  hipoteca__deposit   
1          2  59.0       admin._secondary_married  hipoteca__deposit   
2          3  56.0       admin._secondary_married            deposit   
3          4  41.0   technician_secondary_married  hipoteca__deposit   
4          5  55.0     services_secondary_married  hipoteca__deposit   
...      ...   ...                            ...                ...   
16158  21379  80.0      retired_secondary_married            deposit   
16159  21380  30.0  unemployed_secondary_divorced            deposit   
16160  21381  33.0  entrepreneur_secondary_single            deposit   
16161  21382  57.0     technician_primary_married  hipoteca__deposit   
16162  21383  28.0     management_tertiary_single  hipoteca__deposit   

       is_in_default  balance  
0                  0     2343  
1                  0     2343  
2                  0       45  
3                  0     1270  
4                  0     2476  
...              ...      ...  
16158              0     8304  
16159              0     7105  
16160              0      278  
16161              0     3376  
16162              0       41  

[13079 rows x 6 columns]

In [28]:
tabla_contingencia = pd.crosstab(cliente_selec['perfil'], cliente_selec['producto'])
tabla_contingencia = tabla_contingencia.sort_values(by='deposit', ascending = False).head(50)
tabla_contingencia

producto                        credito  credito_deposit  deposit  hipoteca  \
perfil                                                                        
management_tertiary_married          65               26      546       386   
management_tertiary_single           29               14      421       197   
retired_secondary_married            12                6      209        33   
student_secondary_single              0                0      202        20   
admin._secondary_married             33                5      196       226   
technician_secondary_married         31               29      194       224   
retired_primary_married               7                1      175        21   
technician_secondary_single          29               23      135       147   
technician_tertiary_single            4                1      131        72   
technician_tertiary_married           7                6      115        52   
admin._secondary_single              26               17      113       150   
management_tertiary_divorced         13               14      101        85   
blue-collar_secondary_married        34               21      101       414   
services_secondary_married           21               14       85       228   
blue-collar_primary_married          46                7       80       299   
services_secondary_single            23                8       79       111   
retired_tertiary_married              6                0       76         9   
student_tertiary_single               2                0       71        15   
blue-collar_secondary_single         16               24       69       151   
self-employed_tertiary_single         2                1       69        18   
unemployed_secondary_married          1                0       67        42   
retired_primary_divorced              1                2       59         3   
student_unknown_single                0                0       56         2   
management_secondary_married          4                5       54        60   
self-employed_tertiary_married        6                0       51        22   
retired_secondary_divorced            2                3       49         6   
retired_unknown_married               1                0       45         1   
admin._secondary_divorced            14               10       44        64   
unemployed_tertiary_single            0                0       41         8   
admin._tertiary_married               2                5       40        16   
technician_secondary_divorced        15                3       36        59   
management_unknown_married            0                0       36        11   
housemaid_primary_married             9                5       33        30   
unemployed_secondary_single           0                2       32        15   
entrepreneur_tertiary_married        10                1       29        35   
admin._tertiary_single                4                6       28        29   
retired_tertiary_divorced             0                2       26         4   
entrepreneur_secondary_married       11                8       25        35   
unemployed_secondary_divorced         2                0       25        14   
blue-collar_tertiary_single           2                0       25         8   
student_primary_single                0                0       23         0   
unemployed_tertiary_married           1                0       22         2   
services_secondary_divorced           6                6       21        36   
housemaid_secondary_married           0                4       20        13   
management_primary_married            4                0       19        22   
housemaid_primary_divorced            0                0       16         3   
blue-collar_primary_single            7                1       16        59   
blue-collar_primary_divorced          8                3       16        24   
services_tertiary_single              0           

In [29]:
# Variable categórica y variable categórica: perfil y producto
# Example of the Chi-Squared Test
from scipy.stats import chi2_contingency
stat, p, dof, expected = chi2_contingency(tabla_contingencia)
print('stat=%.3f, p=%.3f' % (stat, p))
if p > 0.05:
	print('Probably independent')
else:
	print('Probably dependent')

## Según el resultado del test, existe una relación entre los diferentes perfiles y los productos contratados

stat=2647.475, p=0.000
Probably dependent


## OFERTAS: estas campañas captan a más clientes y fidelizan a los que ya tienen
1) Openbank y Caja Rural Central: “Ambas cuentan con un plan amigo con el que un cliente de la entidad puede ganar unos pocos cientos de euros si un conocido suyo contrata una hipoteca con un código promocional”.
2) el regalo de vajillas o toallas en verano en tiempos ya pasados, pasando por portátiles, patinetes eléctricos, smartphones o wearables, los bancos siempre han tenido los regalos como una parte fundamental de su estrategia de fidelización de clientes.
3) las recompensas desde wearables hasta suscripciones gratuitas en plataformas de streaming como Netflix, HBO o Amazon Prime.
4) Openbank. Su nueva campaña, ‘Invita a un amigo Hipoteca Open’, regala hasta 1.500 euros por invitar a un máximo de diez amigos o familiares a contratar o traer su hipoteca al banco. Además, los nuevos usuarios obtienen un abono de 300 euros si se dan de alta en la oferta antes del 9 de septiembre y contratar o traer una hipoteca a Openbank hasta el 31 de octubre
5) Caja Rural Central “regala” 300 euros en la tarjeta de crédito por invitar a tus amigos a contratar una hipoteca.
6) Los bancos ahora “regalan” más dinero con la domiciliación de nóminas. Openbank es, otra vez, una de las entidades que se ha sumado a premiar las cuentas de los clientes. La entidad ofrecía en mayo 40 euros a nuevos clientes que contratasen la Cuenta Corriente Open e hiciesen un ingreso inicial de al menos 300 euros que deberán mantener hasta que se realice el abono del incentivo durante el mes de julio.
7) Otro ejemplo de esta estrategia es Targobank, que también abona 40 euros por hacerse una cuenta. En Unicaja, al domiciliar la nómina te dan una tarjeta de 150 euros de El Corte Inglés. Abanca, por su parte, abona 300 euros al domiciliar la nómina si se hace antes del 31 de diciembre.
8) BBVA hay variedad de recompensas. La entidad da 150 euros al domiciliar la nómina con la campaña “BBVA va contigo”. Además, tiene una cuenta online sin comisiones con la que te reembolsan una de estas suscripciones; Netflix, Spotify, HBO Max, Disney Plus, PlayStation Plus o DAZN durante 6 meses (hasta 9,99 euros al mes) si se paga con la tarjeta Aqua Débito.
9) aquellos jóvenes en edades comprendidas entre 18 y 29 años que deseen hacerse clientes de BBVA en España, podrán disfrutar gratuitamente de los servicios de Uber y recibir reembolsos por sus pagos para apoyar a sus ‘streamers’ favoritos en Twitch.
10) los mantenimientos de cuenta de ahorro gratuitos sin mantienes un saldo alto durante un año
11) Ventas cruzadas:

#### Los productos de inversión más importantes son: 
#### Fondos de inversión: 
la cantidad de fondos existentes, la obtención de rentabilidad desde pequeñas cantidades, la seguridad al estar gestionados por entidades y profesionales, y la posibilidad de poder retirar tu dinero en cualquier momento son algunas características que los han convertido en uno de los productos más demandados. 
#### Acciones: 
las acciones son una forma de participar en el capital social de las empresas. Si el valor de una compañía crece, sus acciones también lo hacen, pudiendo obtener ganancias vendiendo tus participaciones. Si el valor decrece, sus acciones caen y sufrirás pérdidas. 
#### Bonos: 
se trata de comprar deuda pública emitida por el Estado o privada por las empresas para conseguir financiación. Finalizado un período de tiempo determinado, recibes el dinero invertido más la rentabilidad que se haya generado. 

 ### tipos de programas de fidelización de clientes que tu empresa puede aprovechar:
- Programas de puntos
- Programas por niveles
- Programas basados en misiones
- Programas basados en el gasto
- Programas de juego
- Programas de ventajas gratuitas
- Programas de suscripción
- Programas de la comunidad
- Programas de recomendación de amigos
- Programas de pago
- Programas de devolución de dinero

#### AGRUPACION POR EDAD MINIMA, MEDIA y MAXIMA, BALANCE MINIMO, MEDIO Y MAXIMO, Y RECUENTO DE CLIENTES POR PERFIL Y PRODUCTO

In [30]:
cliente_resumen = cliente_selec.groupby(['perfil', 'producto']).agg(edad_min=('age','min'),edad_media=('age','mean'),edad_max=('age','max'), balance_min=('balance','min'),balance_medio=('balance', 'mean'), balance_max=('balance','max'),num_cliente=('id','count')).reset_index()
cliente_resumen['edad_media'] = cliente_resumen['edad_media'].round(0)
cliente_resumen['balance_medio'] = cliente_resumen['balance_medio'].round(1)
cliente_resumen
cliente_resumen_mayor = cliente_resumen.sort_values('num_cliente', ascending=False, ignore_index = True).reset_index()
cliente_resumen_mayor

index                         perfil          producto  edad_min  \
0        0    management_tertiary_married           deposit      25.0   
1        1     management_tertiary_single           deposit      23.0   
2        2  blue-collar_secondary_married          hipoteca      23.0   
3        3    management_tertiary_married          hipoteca      27.0   
4        4    blue-collar_primary_married          hipoteca      26.0   
..     ...                            ...               ...       ...   
552    552         admin._tertiary_single  hipoteca_credito      32.0   
553    553      unemployed_unknown_single           deposit      29.0   
554    554         admin._primary_married           deposit      72.0   
555    555       unknown_unknown_divorced           deposit      45.0   
556    556        admin._primary_divorced           credito      56.0   

     edad_media  edad_max  balance_min  balance_medio  balance_max  \
0          44.0      73.0         -375         2445.6        31472   
1          33.0      61.0          -27         2345.1        20585   
2          39.0      60.0         -972         1088.2        19447   
3          41.0      60.0         -700         1465.1        23189   
4          41.0      60.0         -848         1076.8        15841   
..          ...       ...          ...            ...          ...   
552        32.0      32.0          328          328.0          328   
553        29.0      29.0         1584         1584.0         1584   
554        72.0      72.0         2321         2321.0         2321   
555        45.0      45.0            0            0.0            0   
556        56.0      56.0            0            0.0            0   

     num_cliente  
0            546  
1            421  
2            414  
3            386  
4            299  
..           ...  
552            1  
553            1  
554            1  
555            1  
556            1  

[557 rows x 10 columns]

### HIPOTECA CON VALORES ORDENADOS DE MAYOR A MENOR

In [31]:
data1 = cliente_resumen_mayor[cliente_resumen_mayor['producto']== 'hipoteca'].reset_index()
data1 = data1.rename(columns = {'num_cliente':'hipoteca'})
data1 = data1.drop(columns=['level_0','producto'])
data1.head(10)

index                         perfil  edad_min  edad_media  edad_max  \
0      2  blue-collar_secondary_married      23.0        39.0      60.0   
1      3    management_tertiary_married      27.0        41.0      60.0   
2      4    blue-collar_primary_married      26.0        41.0      60.0   
3      5     services_secondary_married      23.0        38.0      60.0   
4      6       admin._secondary_married      23.0        39.0      60.0   
5      7   technician_secondary_married      25.0        40.0      63.0   
6     13     management_tertiary_single      23.0        34.0      59.0   
7     18   blue-collar_secondary_single      23.0        32.0      56.0   
8     19        admin._secondary_single      22.0        35.0      56.0   
9     20    technician_secondary_single      23.0        35.0      56.0   

   balance_min  balance_medio  balance_max  hipoteca  
0         -972         1088.2        19447       414  
1         -700         1465.1        23189       386  
2         -848         1076.8        15841       299  
3         -522          903.1         9374       228  
4         -666          688.5         8784       226  
5        -1701         1480.1        29207       224  
6         -508         1340.6        10269       197  
7         -575          999.1        13156       151  
8         -498         1254.5        10218       150  
9         -397         1277.3        18777       147

### DEPOSIT CON VALORES ORDENADOS DE MAYOR A MENOR

In [32]:
data2 = cliente_resumen_mayor[cliente_resumen_mayor['producto']== 'deposit'].reset_index()
data2 = data2.rename(columns = {'num_cliente':'deposit'})
data2 = data2.drop(columns=['level_0','producto'])
data2.head(10)

index                        perfil  edad_min  edad_media  edad_max  \
0      0   management_tertiary_married      25.0        44.0      73.0   
1      1    management_tertiary_single      23.0        33.0      61.0   
2      8     retired_secondary_married      43.0        67.0      84.0   
3     11      student_secondary_single      19.0        25.0      36.0   
4     14      admin._secondary_married      26.0        46.0      66.0   
5     15  technician_secondary_married      26.0        45.0      71.0   
6     16       retired_primary_married      34.0        70.0      87.0   
7     21   technician_secondary_single      23.0        32.0      53.0   
8     22    technician_tertiary_single      25.0        33.0      54.0   
9     24   technician_tertiary_married      25.0        42.0      71.0   

   balance_min  balance_medio  balance_max  deposit  
0         -375         2445.6        31472      546  
1          -27         2345.1        20585      421  
2            0         3815.9        81204      209  
3            0         1262.1        23878      202  
4            0         1860.5        14968      196  
5         -308         1879.9        15485      194  
6            0         2898.3        29340      175  
7            0         1797.5        18508      135  
8            0         2931.2        34646      131  
9         -215         1735.9        23076      115

### HIPOTECA + DEPOSITO CON VALORES ORDENADOS DE MAYOR A MENOR

In [33]:
data3 = cliente_resumen_mayor[cliente_resumen_mayor['producto']== 'hipoteca__deposit'].reset_index()
data3 = data3.rename(columns = {'num_cliente':'hipoteca__deposit'})
data3 = data3.drop(columns=['level_0','producto'])
data3.head(10)

index                         perfil  edad_min  edad_media  edad_max  \
0      9  blue-collar_secondary_married      24.0        39.0      58.0   
1     10    management_tertiary_married      28.0        39.0      61.0   
2     12     management_tertiary_single      24.0        33.0      60.0   
3     17   technician_secondary_married      24.0        40.0      63.0   
4     23       admin._secondary_married      27.0        41.0      63.0   
5     28    blue-collar_primary_married      24.0        42.0      59.0   
6     29    technician_secondary_single      22.0        33.0      49.0   
7     31        admin._secondary_single      25.0        36.0      57.0   
8     33   blue-collar_secondary_single      20.0        31.0      45.0   
9     34     services_secondary_married      24.0        38.0      59.0   

   balance_min  balance_medio  balance_max  hipoteca__deposit  
0         -519         1263.2        10287                208  
1         -970         1564.7        22867                205  
2         -713         2211.1        36252                200  
3        -1944         1407.2        10685                153  
4         -522         1383.9        10250                127  
5         -522         1454.4        12519                105  
6        -1129         1297.2         9131                105  
7         -930         1273.9        18188                101  
8         -416          852.9         7773                101  
9         -538         1792.4        11650                 90

### HIPOTECA+DEPOSIT+CREDITO CON VALORES ORDENADOS DE MAYOR A MENOR

In [34]:
data4 = cliente_resumen_mayor[cliente_resumen_mayor['producto']== 'hipoteca_credito_deposit'].reset_index()
data4 = data4.rename(columns = {'num_cliente':'_3productos'})
data4 = data4.drop(columns=['level_0','producto'])
data4.head(12)

index                         perfil  edad_min  edad_media  edad_max  \
0      65    management_tertiary_married      25.0        41.0      59.0   
1      77  blue-collar_secondary_married      24.0        36.0      56.0   
2     105       admin._secondary_married      26.0        38.0      58.0   
3     126    blue-collar_primary_married      21.0        39.0      60.0   
4     127   blue-collar_secondary_single      24.0        34.0      58.0   
5     131   technician_secondary_married      27.0        45.0      59.0   
6     133        admin._secondary_single      27.0        38.0      59.0   
7     136     services_secondary_married      26.0        35.0      42.0   
8     162      admin._secondary_divorced      28.0        43.0      60.0   
9     186    technician_secondary_single      27.0        36.0      49.0   
10    203     management_tertiary_single      25.0        32.0      42.0   
11    239    technician_tertiary_married      30.0        43.0      60.0   

    balance_min  balance_medio  balance_max  _3productos  
0          -556         1069.7         4119           47  
1          -839          331.9         1598           37  
2          -106         1504.8         8749           26  
3          -537          107.3         1262           21  
4             2          887.7         3043           21  
5          -754          345.6         2183           21  
6          -528          357.4         2263           20  
7          -379          969.2         4151           19  
8          -504         1390.4         6046           15  
9          -522          -13.3          265           12  
10            0          456.1         1618           11  
11          -71         1328.2         4654            8

### HIPOTECA + CREDITO CON VALORES ORDENADOS DE MAYOR A MENOR

In [43]:
data5 = cliente_resumen_mayor[cliente_resumen_mayor['producto']== 'hipoteca_credito'].reset_index()
data5 = data5.rename(columns = {'num_cliente':'hipo_loan'})
data5 = data5.drop(columns=['level_0','producto'])
data5.head(50)

index                           perfil  edad_min  edad_media  edad_max  \
0      25    blue-collar_secondary_married      26.0        38.0      60.0   
1      44     technician_secondary_married      29.0        41.0      57.0   
2      49      management_tertiary_married      29.0        41.0      56.0   
3      51         admin._secondary_married      28.0        42.0      58.0   
4      52       services_secondary_married      27.0        39.0      59.0   
5      57      blue-collar_primary_married      28.0        40.0      60.0   
6      96          admin._secondary_single      24.0        36.0      59.0   
7     102      technician_secondary_single      23.0        32.0      54.0   
8     114        admin._secondary_divorced      29.0        44.0      59.0   
9     122       management_tertiary_single      29.0        36.0      56.0   
10    137      technician_tertiary_married      26.0        37.0      52.0   
11    144      services_secondary_divorced      34.0        44.0      56.0   
12    148    technician_secondary_divorced      29.0        42.0      58.0   
13    163        services_secondary_single      24.0        31.0      45.0   
14    165     entrepreneur_primary_married      37.0        48.0      58.0   
15    181   entrepreneur_secondary_married      27.0        38.0      57.0   
16    199     unemployed_secondary_married      28.0        34.0      42.0   
17    200   self-employed_tertiary_married      28.0        43.0      52.0   
18    201     management_tertiary_divorced      31.0        43.0      51.0   
19    214  self-employed_secondary_married      26.0        37.0      51.0   
20    217          admin._tertiary_married      28.0        31.0      35.0   
21    219     blue-collar_secondary_single      26.0        31.0      38.0   
22    222    entrepreneur_tertiary_married      32.0        41.0      55.0   
23    227   blue-collar_secondary_divorced      32.0        42.0      51.0   
24    235          retired_primary_married      54.0        58.0      61.0   
25    247     management_secondary_married      32.0        38.0      52.0   
26    272        housemaid_primary_married      44.0        50.0      58.0   
27    277       technician_unknown_married      38.0        49.0      55.0   
28    278       technician_tertiary_single      28.0        35.0      42.0   
29    280        retired_secondary_married      54.0        57.0      60.0   
30    289       management_primary_married      39.0        46.0      57.0   
31    309       blue-collar_primary_single      28.0        37.0      44.0   
32    361  self-employed_tertiary_divorced      45.0        48.0      54.0   
33    370   entrepreneur_tertiary_divorced      45.0        46.0      48.0   
34    371     blue-collar_primary_divorced      40.0        44.0      49.0   
35    373        retired_tertiary_divorced      54.0        57.0      58.0   
36    377       technician_primary_married      32.0        33.0      36.0   
37    379           admin._primary_married      42.0        47.0      57.0   
38    388      blue-collar_unknown_married      34.0        41.0      56.0   
39    393         services_primary_married      33.0        41.0      45.0   
40    399    entrepreneur_secondary_single      30.0        30.0      30.0   
41    406    self-employed_tertiary_single      39.0        39.0      39.0   
42    410       unemployed_primary_married      43.0        43.0      43.0   
43    414   self-employed_secondary_single      26.0        26.0      26.0   
44    417      housemaid_secondary_married      27.0        34.0      40.0   
45    424      unemployed_secondary_single      45.0        45.0      45.0   
46    450  entrepreneur_secondary_divorced      41.0        44.0      47.0   
47    456         retired_primary_divorced      60.0        60.0      60.0   
48    457          services_primary_single      26.0        26.0      26.0   
49    464       management_unknown_married      59.0        59.0      59.0   

    balance_min  balance_